dns-320l_GPL/shadow-4.0.18.2/man/pwck.8
.\"     Title: pwck
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.2 <http://docbook.sf.net/>
.\"      Date: 10/28/2007
.\"    Manual: System Management Commands
.\"    Source: System Management Commands
.\"
.TH "PWCK" "8" "10/28/2007" "System Management Commands" "System Management Commands"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
pwck - verify integrity of password files
.SH "SYNOPSIS"
.HP 5
\fBpwck\fR [\-q] [\-s] [\fIpasswd\fR\ \fIshadow\fR]
.HP 5
\fBpwck\fR [\-q] [\-r] [\fIpasswd\fR\ \fIshadow\fR]
.SH "DESCRIPTION"
.PP
The
\fBpwck\fR
command verifies the integrity of the system authentication information\. All entries in the
\fI/etc/passwd\fR
and
\fI/etc/shadow\fR
are checked to see that the entry has the proper format and valid data in each field\. The user is prompted to delete entries that are improperly formatted or which have other uncorrectable errors\.
.PP
Checks are made to verify that each entry has:
.sp
.RS 4
\h'-04'\(bu\h'+03'the correct number of fields
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'a unique user name
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'a valid user and group identifier
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'a valid primary group
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'a valid home directory
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'a valid login shell
.RE
.PP
The checks for correct number of fields and unique user name are fatal\. If the entry has the wrong number of fields, the user will be prompted to delete the entire line\. If the user does not answer affirmatively, all further checks are bypassed\. An entry with a duplicated user name is prompted for deletion, but the remaining checks will still be made\. All other errors are warning and the user is encouraged to run the
\fBusermod\fR
command to correct the error\.
.PP
The commands which operate on the
\fI/etc/passwd\fR
file are not able to alter corrupted or duplicated entries\.
\fBpwck\fR
should be used in those circumstances to remove the offending entry\.
.SH "OPTIONS"
.PP
The options which apply to the
\fBpwck\fR
command are:
.PP
\fB\-q\fR
.RS 4
Report errors only\. The warnings which do not require any action from the user won\'t be displayed\.
.RE
.PP
\fB\-r\fR
.RS 4
Execute the
\fBpwck\fR
command in read\-only mode\.
.RE
.PP
\fB\-s\fR
.RS 4
Sort entries in
\fI/etc/passwd\fR
and
\fI/etc/shadow\fR
by UID\.
.RE
.PP
By default,
\fBpwck\fR
operates on the files
\fI/etc/passwd\fR
and
\fI/etc/shadow\fR\. The user may select alternate files with the
\fIpasswd\fR
and
\fIshadow\fR
parameters\.
.SH "FILES"
.PP
\fI/etc/group\fR
.RS 4
Group account information\.
.RE
.PP
\fI/etc/passwd\fR
.RS 4
User account information\.
.RE
.PP
\fI/etc/shadow\fR
.RS 4
Secure user account information\.
.RE
.SH "SEE ALSO"
.PP
\fBgroup\fR(5),
\fBpasswd\fR(5),
\fBshadow\fR(5),
\fBusermod\fR(8)\.
.SH "EXIT VALUES"
.PP
The
\fBpwck\fR
command exits with the following values:
.PP
\fI0\fR
.RS 4
success
.RE
.PP
\fI1\fR
.RS 4
invalid command syntax
.RE
.PP
\fI2\fR
.RS 4
one or more bad password entries
.RE
.PP
\fI3\fR
.RS 4
can\'t open password files
.RE
.PP
\fI4\fR
.RS 4
can\'t lock password files
.RE
.PP
\fI5\fR
.RS 4
can\'t update password files
.RE


dns-320l_GPL/shadow-4.0.18.2/src/pwck.c
/*
 * Copyright 1992 - 1994, Julianne Frances Haugh
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Julianne F. Haugh nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <config.h>

#ident "$Id: pwck.c,v 1.1.1.1 2009/10/09 02:53:08 jack Exp $"

#include <fcntl.h>
#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include "chkname.h"
#include "commonio.h"
#include "defines.h"
#include "prototypes.h"
#include "pwio.h"
#include "shadowio.h"
#include "getdef.h"
#include "nscd.h"
extern void __pw_del_entry (const struct commonio_entry *);
extern struct commonio_entry *__pw_get_head (void);

extern void __spw_del_entry (const struct commonio_entry *);
extern struct commonio_entry *__spw_get_head (void);

/*
 * Exit codes
 */

#define	E_OKAY		0
#define	E_USAGE		1
#define	E_BADENTRY	2
#define	E_CANTOPEN	3
#define	E_CANTLOCK	4
#define	E_CANTUPDATE	5

/*
 * Global variables
 */

static char *Prog;
static const char *pwd_file = PASSWD_FILE;

static const char *spw_file = SHADOW_FILE;
static int read_only = 0;
static int quiet = 0;		/* don't report warnings, only errors */

/* local function prototypes */
static void usage (void);
static int yes_or_no (void);

/*
 * usage - print syntax message and exit
 */
static void usage (void)
{
	fprintf (stderr, _("Usage: %s [-q] [-r] [-s] [passwd [shadow]]\n"),
		 Prog);
	exit (E_USAGE);
}

/*
 * yes_or_no - get answer to question from the user
 */
static int yes_or_no (void)
{
	char buf[80];

	/*
	 * In read-only mode all questions are answered "no".
	 */

	if (read_only) {
		printf (_("No\n"));
		return 0;
	}

	/*
	 * Get a line and see what the first character is.
	 */
	if (fgets (buf, sizeof buf, stdin))
		return buf[0] == 'y' || buf[0] == 'Y';

	return 0;
}

/*
 * pwck - verify password file integrity
 */
int main (int argc, char **argv)
{
	int arg;
	int errors = 0;
	int changed = 0;
	struct commonio_entry *pfe, *tpfe;
	struct passwd *pwd;
	int sort_mode = 0;

	struct commonio_entry *spe, *tspe;
	struct spwd *spw;
	int is_shadow = 0;

	/*
	 * Get my name so that I can use it to report errors.
	 */
	Prog = Basename (argv[0]);

	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);

	OPENLOG ("pwck");

	/*
	 * Parse the command line arguments
	 */
	while ((arg = getopt (argc, argv, "eqrs")) != EOF) {
		switch (arg) {
		case 'e':	/* added for Debian shadow-961025-2 compatibility */
		case 'q':
			quiet = 1;
			break;
		case 'r':
			read_only = 1;
			break;
		case 's':
			sort_mode = 1;
			break;
		default:
			usage ();
		}
	}

	if (sort_mode && read_only) {
		fprintf (stderr, _("%s: -s and -r are incompatibile\n"), Prog);
		exit (E_USAGE);
	}

	/*
	 * Make certain we have the right number of arguments
	 */
	if (optind != argc && optind + 1 != argc && optind + 2 != argc)
		usage ();

	/*
	 * If there are two left over filenames, use those as the password
	 * and shadow password filenames.
	 */
	if (optind != argc) {
		pwd_file = argv[optind];
		pw_name (pwd_file);
	}
	if (optind + 2 == argc) {
		spw_file = argv[optind + 1];
		spw_name (spw_file);
		is_shadow = 1;
	} else if (optind == argc)
		is_shadow = spw_file_present ();

	/*
	 * Lock the files if we aren't in "read-only" mode
	 */
	if (!read_only) {
		if (!pw_lock ()) {
			fprintf (stderr, _("%s: cannot lock file %s\n"),
				 Prog, pwd_file);
			if (optind == argc)
				SYSLOG ((LOG_WARN, "cannot lock %s", pwd_file));
			closelog ();
			exit (E_CANTLOCK);
		}
		if (is_shadow && !spw_lock ()) {
			fprintf (stderr, _("%s: cannot lock file %s\n"),
				 Prog, spw_file);
			if (optind == argc)
				SYSLOG ((LOG_WARN, "cannot lock %s", spw_file));
			closelog ();
			exit (E_CANTLOCK);
		}
	}

	/*
	 * Open the files. Use O_RDONLY if we are in read_only mode, O_RDWR
	 * otherwise.
	 */
	if (!pw_open (read_only ? O_RDONLY : O_RDWR)) {
		fprintf (stderr, _("%s: cannot open file %s\n"),
			 Prog, pwd_file);
		if (optind == argc)
			SYSLOG ((LOG_WARN, "cannot open %s", pwd_file));
		closelog ();
		exit (E_CANTOPEN);
	}
	if (is_shadow && !spw_open (read_only ? O_RDONLY : O_RDWR)) {
		fprintf (stderr, _("%s: cannot open file %s\n"),
			 Prog, spw_file);
		if (optind == argc)
			SYSLOG ((LOG_WARN, "cannot open %s", spw_file));
		closelog ();
		exit (E_CANTOPEN);
	}

	if (sort_mode) {
		pw_sort ();
		if (is_shadow)
			spw_sort ();
		goto write_and_bye;
	}

	/*
	 * Loop through the entire password file.
	 */
	for (pfe = __pw_get_head (); pfe; pfe = pfe->next) {
		/*
		 * If this is a NIS line, skip it. You can't "know" what NIS
		 * is going to do without directly asking NIS ...
		 */
		if (pfe->line[0] == '+' || pfe->line[0] == '-')
			continue;

		/*
		 * Start with the entries that are completely corrupt.  They
		 * have no (struct passwd) entry because they couldn't be
		 * parsed properly.
		 */
		if (!pfe->eptr) {
			/*
			 * Tell the user this entire line is bogus and ask
			 * them to delete it.
			 */
			printf (_("invalid password file entry\n"));
			printf (_("delete line '%s'? "), pfe->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */
			if (!yes_or_no ())
				continue;

			/*
			 * All password file deletions wind up here. This
			 * code removes the current entry from the linked
			 * list. When done, it skips back to the top of the
			 * loop to try out the next list element.
			 */
		      delete_pw:
			SYSLOG ((LOG_INFO, "delete passwd line `%s'",
				 pfe->line));
			changed++;

			__pw_del_entry (pfe);
			continue;
		}

		/*
		 * Password structure is good, start using it.
		 */
		pwd = pfe->eptr;

		/*
		 * Make sure this entry has a unique name.
		 */
		for (tpfe = __pw_get_head (); tpfe; tpfe = tpfe->next) {
			const struct passwd *ent = tpfe->eptr;

			/*
			 * Don't check this entry
			 */
			if (tpfe == pfe)
				continue;

			/*
			 * Don't check invalid entries.
			 */
			if (!ent)
				continue;

			if (strcmp (pwd->pw_name, ent->pw_name) != 0)
				continue;

			/*
			 * Tell the user this entry is a duplicate of
			 * another and ask them to delete it.
			 */
			printf (_("duplicate password entry\n"));
			printf (_("delete line '%s'? "), pfe->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */
			if (yes_or_no ())
				goto delete_pw;
		}

		/*
		 * Check for invalid usernames.  --marekm
		 */
		if (!check_user_name (pwd->pw_name)) {
			printf (_("invalid user name '%s'\n"), pwd->pw_name);
			errors++;
		}

		/*
		 * Make sure the primary group exists
		 */
		if (!quiet && !getgrgid (pwd->pw_gid)) {

			/*
			 * No primary group, just give a warning
			 */

			printf (_("user %s: no group %u\n"),
				pwd->pw_name, pwd->pw_gid);
			errors++;
		}

		/*
		 * Make sure the home directory exists
		 */
		if (!quiet && access (pwd->pw_dir, F_OK)) {
			/*
			 * Home directory doesn't exist, give a warning
			 */
			printf (_
				("user %s: directory %s does not exist\n"),
				pwd->pw_name, pwd->pw_dir);
			errors++;
		}

		/*
		 * Make sure the login shell is executable
		 */
		if (!quiet && pwd->pw_shell[0]
		    && access (pwd->pw_shell, F_OK)) {

			/*
			 * Login shell doesn't exist, give a warning
			 */
			printf (_("user %s: program %s does not exist\n"),
				pwd->pw_name, pwd->pw_shell);
			errors++;
		}
#ifdef SHADOWPWD
		/*
		 * Make sure this entry exists in the /etc/gshadow file.
		 */

		if (is_shadow) {
			spw = (struct spwd *) spw_locate (pwd->pw_name);
			if (spw == NULL) {
				printf (_
					("no matching password file entry in %s\n"),
					spw_file);
				printf (_("add user '%s' in %s? "),
					pwd->pw_name, spw_file);
				errors++;
				if (yes_or_no ()) {
					struct spwd sp;
					struct passwd pw;

					sp.sp_namp = pwd->pw_name;
					sp.sp_pwdp = pwd->pw_passwd;
					sp.sp_min =
					    getdef_num ("PASS_MIN_DAYS", -1);
					sp.sp_max =
					    getdef_num ("PASS_MAX_DAYS", -1);
					sp.sp_warn =
					    getdef_num ("PASS_WARN_AGE", -1);
					sp.sp_inact = -1;
					sp.sp_expire = -1;
					sp.sp_flag = -1;
					sp.sp_lstchg =
					    time ((time_t *) 0) / (24L * 3600L);
					changed++;

					if (!spw_update (&sp)) {
						fprintf (stderr,
							 _
							 ("%s: can't update shadow entry for %s\n"),
							 Prog, sp.sp_namp);
						exit (E_CANTUPDATE);
					}
					/* remove password from /etc/passwd */
					pw = *pwd;
					pw.pw_passwd = SHADOW_PASSWD_STRING;	/* XXX warning: const */
					if (!pw_update (&pw)) {
						fprintf (stderr,
							 _
							 ("%s: can't update passwd entry for %s\n"),
							 Prog, pw.pw_name);
						exit (E_CANTUPDATE);
					}
				}
			}
		}
#endif
	}

	if (!is_shadow)
		goto shadow_done;

	/*
	 * Loop through the entire shadow password file.
	 */
	for (spe = __spw_get_head (); spe; spe = spe->next) {
		/*
		 * Do not treat lines which were missing in gshadow
		 * and were added earlier.
		 */
		if (spe->line == NULL)
			continue;

		/*
		 * If this is a NIS line, skip it. You can't "know" what NIS
		 * is going to do without directly asking NIS ...
		 */
		if (spe->line[0] == '+' || spe->line[0] == '-')
			continue;

		/*
		 * Start with the entries that are completely corrupt. They
		 * have no (struct spwd) entry because they couldn't be
		 * parsed properly.
		 */
		if (!spe->eptr) {
			/*
			 * Tell the user this entire line is bogus and ask
			 * them to delete it.
			 */
			printf (_("invalid shadow password file entry\n"));
			printf (_("delete line '%s'? "), spe->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */
			if (!yes_or_no ())
				continue;

			/*
			 * All shadow file deletions wind up here. This code
			 * removes the current entry from the linked list.
			 * When done, it skips back to the top of the loop
			 * to try out the next list element.
			 */
		      delete_spw:
			SYSLOG ((LOG_INFO, "delete shadow line `%s'",
				 spe->line));
			changed++;

			__spw_del_entry (spe);
			continue;
		}

		/*
		 * Shadow password structure is good, start using it.
		 */
		spw = spe->eptr;

		/*
		 * Make sure this entry has a unique name.
		 */
		for (tspe = __spw_get_head (); tspe; tspe = tspe->next) {
			const struct spwd *ent = tspe->eptr;

			/*
			 * Don't check this entry
			 */
			if (tspe == spe)
				continue;

			/*
			 * Don't check invalid entries.
			 */
			if (!ent)
				continue;

			if (strcmp (spw->sp_namp, ent->sp_namp) != 0)
				continue;

			/*
			 * Tell the user this entry is a duplicate of
			 * another and ask them to delete it.
			 */
			printf (_("duplicate shadow password entry\n"));
			printf (_("delete line '%s'? "), spe->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */
			if (yes_or_no ())
				goto delete_spw;
		}

		/*
		 * Make sure this entry exists in the /etc/passwd
		 * file.
		 */
		if (!pw_locate (spw->sp_namp)) {
			/*
			 * Tell the user this entry has no matching
			 * /etc/passwd entry and ask them to delete it.
			 */
			printf (_("no matching password file entry in %s\n"),
				pwd_file);
			printf (_("delete line '%s'? "), spe->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */
			if (yes_or_no ())
				goto delete_spw;
		}

		/*
		 * Warn if last password change in the future.  --marekm
		 */
		if (!quiet && spw->sp_lstchg > time ((time_t *) 0) / SCALE) {
			printf (_
				("user %s: last password change in the future\n"),
				spw->sp_namp);
			errors++;
		}
	}

      shadow_done:

	/*
	 * All done. If there were no change we can just abandon any
	 * changes to the files.
	 */
	if (changed) {
	      write_and_bye:
		if (!pw_close ()) {
			fprintf (stderr, _("%s: cannot update file %s\n"),
				 Prog, pwd_file);
			SYSLOG ((LOG_WARN, "cannot update %s", pwd_file));
			closelog ();
			exit (E_CANTUPDATE);
		}
		if (is_shadow && !spw_close ()) {
			fprintf (stderr, _("%s: cannot update file %s\n"),
				 Prog, spw_file);
			SYSLOG ((LOG_WARN, "cannot update %s", spw_file));
			closelog ();
			exit (E_CANTUPDATE);
		}
	}

	/*
	 * Don't be anti-social - unlock the files when you're done.
	 */
	if (is_shadow)
		spw_unlock ();
	(void) pw_unlock ();

	nscd_flush_cache ("passwd");

	/*
	 * Tell the user what we did and exit.
	 */
	if (errors)
		printf (changed ?
			_("%s: the files have been updated\n") :
			_("%s: no changes\n"), Prog);

	closelog ();
	exit (errors ? E_BADENTRY : E_OKAY);
}
